File: algorithm/unique-binary-search-trees.cpp
96. Unique Binary Search Trees
Difficulty: Medium

Given n, how many structurally unique BST's (binary search trees) that store values 1...n?

For example,
Given n = 3, there are a total of 5 unique BST's.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3

// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int numTrees(int n) {
        vector<int> f(n + 1);
        f[0] = f[1] = 1;
        for (int i = 2; i <= n; ++i) {
            for (int k = 1; k <= i; ++k) {
                f[i] += f[i - k] * f[k - 1];
            } 
        }
        return f[n];
    }
};

File: data-structure-and-algorithm/java/unique-binary-search-trees.java
/**
 * Unique Binary Search Trees
 * 
 * Given n, how many structurally unique BST's (binary search trees) that store values 1...n?
 * 
 * For example,
 * Given n = 3, there are a total of 5 unique BST's.
 * 
 *    1         3     3      2      1
 *     \       /     /      / \      \
 *      3     2     1      1   3      2
 *     /     /       \                 \
 *    2     1         2                 3
 */

public class Solution {
    public int numTrees(int n) {
        int[] numberOfTrees = new int[n + 1];
        numberOfTrees[0] = numberOfTrees[1] = 1;
        for (int i = 2; i <= n; ++i) {
            for (int j = 0; j < i; ++j) {
                numberOfTrees[i] += numberOfTrees[j] * numberOfTrees[i - j - 1];
            }
        }
        return numberOfTrees[n];
    }
}
